numpy/_core/src/umath/umathmodule.c

```diff
@@ -99,65 +174,119 @@ ufunc_frompyfunc(PyObject *NPY_UNUSED(dummy), PyObject *args, PyObject *kwds) {
 
     if (self == NULL) {
-        PyArray_free(ptr);
+        PyArray_free(str);
+        return NULL;
+    }
+
+    PyArray_DTypeMeta *op_dtypes[NPY_MAXARGS];
+    for (int arg = 0; arg < nargs; arg++) {
+        op_dtypes[arg] = PyArray_DTypeFromTypeNum(NPY_OBJECT);
+        /* These DTypes are immortal and adding INCREFs: so borrow it */
+        Py_DECREF(op_dtypes[arg]);
+    }
+
+    char method_name[101];
+    snprintf(method_name, 100, "ArrayMethod_implementation_for_%s", str);
+
+    NPY_ARRAYMETHOD_FLAGS flags = NPY_METH_REQUIRES_PYAPI | NPY_METH_NO_FLOATINGPOINT_ERRORS;
+
+    PyArrayMethod_GetReductionInitial *get_reduction_intial = NULL;
+    if (nin == 2 && nout == 1) {
+        npy_bool reorderable = NPY_FALSE;
+        PyObject *identity_obj = PyUFunc_GetDefaultIdentity(
+                self, &reorderable);
+        if (identity_obj == NULL) {
+            PyArray_free(str);
+            return NULL;
+        }
+        /*
+         * TODO: For object, "reorderable" is needed(?), because otherwise
+         *       we disable multi-axis reductions `arr.sum(0, 1)`. But for
+         *       `arr = array([["a", "b"], ["c", "d"]], dtype="object")`
+         *       it isn't actually reorderable (order changes result).
+         */
+        if (reorderable) {
+            flags |= NPY_METH_IS_REORDERABLE;
+        }
+        if (identity_obj != Py_None) {
+            get_reduction_intial = &get_initial_from_ufunc;
+        }
+    }
+
+    PyType_Slot slots[4] = {
+        {NPY_METH_strided_loop, &pyfunc_loop},
+        {NPY_METH_get_reduction_initial, get_reduction_intial},
+        {_NPY_METH_static_data, function},
+        {0, NULL},
+    };
+    PyArrayMethod_Spec spec = {
+        .name = method_name,
+        .nin = nin,
+        .nout = nout,
+        .casting = NPY_NO_CASTING,
+        .flags = flags,
+        .dtypes = op_dtypes,
+        .slots = slots,
+    };
+    int res = PyUFunc_AddLoopFromSpec_int((PyObject *)self, &spec, 0);
+    if (res < 0) {
+        PyArray_free(str);
+        return NULL;
+    }
+
+    PyObject *promoter_obj = PyCapsule_New((void *) frompyfunc_promoter, "numpy._ufunc_promoter", NULL);
+    if (promoter_obj == NULL) {
+        PyArray_free(str);
+        return NULL;
+    }
+
+    PyObject *dtypes_tuple = PyTuple_New(nargs);
+    if (dtypes_tuple == NULL) {
+        Py_DECREF(promoter_obj);
+        PyArray_free(str);
         return NULL;
     }
+    for (int i = 0; i < nargs; i++) {
+        PyTuple_SET_ITEM(dtypes_tuple, i, Py_None);
+    }
+
+    PyObject *info = PyTuple_Pack(2, dtypes_tuple, promoter_obj);
+    Py_DECREF(dtypes_tuple);
+    Py_DECREF(promoter_obj);
+    if (info == NULL) {
+        PyArray_free(str);
+        return NULL;
+    }
+
+    res = PyUFunc_AddLoop(self, info, 0);
+    if (res < 0) {
+        PyArray_free(str);
+        return NULL;
+    }
+
     Py_INCREF(function);
     self->obj = function;
-    self->ptr = ptr;
+    self->ptr = str;
```

Comment:
I think we should move this directly after the ufunc is created.  FWIW, the `Py_INCREF(function)` means we explicitly make the function immortal.
(That seems strange, especially on error but I am OK with keeping it for now.)

In principle, rather than `PyArray_free(str)` we should have a `Py_DECREF(self)` on error everywhere.
For a single thing that doesn't matter much, but in C, the pattern:
```
    goto error;

  error:
    Py_DECREF(self);
    return NULL;
```
is nice (and simplifies things a lot if you need to clean up more than one thing).